typing/env.ml

```diff
@@ -2456,6 +2464,37 @@ and lookup_functor_components ~errors ~use ~loc lid env =
   | Error (No_components_alias p) ->
       may_lookup_error errors loc env (Cannot_scrape_alias(lid, p))
 
+and lookup_apply ~errors ~use ~loc lid0 env =
+  let lid_f0, args0 = decompose_apply lid0 in
+  let path_f0, comp_f0 =
+    lookup_module_components ~errors ~use ~loc lid_f0 env in
+  let check_one_apply ~errors ~use ~loc (lid,comp) lid_arg env =
+    let comp_f, mty_param = get_functor_components ~errors ~loc lid env comp in
+    let path_arg, md_arg = lookup_module ~errors ~use ~loc lid_arg env in
+    check_functor_appl
+      ~errors ~loc (lid0, path_f0, args0)
```

Comment:
I don't love that we are passing the original arguments list as long identifiers. This means that we are looking up some of the arguments twice, and I generally think we should aim to throw away long identifiers as soon as they are looked up. This way we avoid the risk of repeating warnings or of accidentally using a different environment and getting inconsistent results.

Since we are clearly checking applications as a whole now, maybe we should lookup all the arguments before checking that the applications are well typed. This probably gives better error messages anyway, and we can keep paths and module types for the arguments list instead of long identifiers.

Reply:
I agree. We wanted to avoid changing the order of typechecking in the normal path, but we could do that.